### ISSUE TITLE
BUG: Disk Mitigation does not work.

### ISSUE THREAD

**Issue by markwragg** [microsoft/service-fabric-healer#146]
Hi,

I'm sure I'm probably doing something wrong here but I can't seem to get a Disk mitigation to fire. I've set `EnableDiskRepair` to true in my `ApplicationManifest.xml` and my `DiskRules.guan` contains the following:

```
Mitigate :- CheckInsideRunInterval(02:00:00), !.

## Added 2023-04-05 | Mark Wragg: Remove files from Service Fabric Observer and Fabric Healer log directories if directories > 1GB

Mitigate(MetricName=?MetricName) :- LogRule(52), match(?MetricName, "DiskSpace"), GetRepairHistory(?repairCount, 08:00:00), 
	?repairCount < 4,
	member(config(?X,?Y), [config("C:\cluster_observer_logs", 1), config("C:\fabric_healer_logs", 1), config("C:\fabric_observer_logs", 1)]), 
	CheckFolderSize(?X, MaxFolderSizeGB=?Y),
	DeleteFiles(?X, SortOrder=Ascending, MaxFilesToDelete=10, RecurseSubdirectories=true).

## Added 2023-04-05 | Mark Wragg: Remove ETL files from Service Fabric Log Traces directory if directory > 20GB

Mitigate(MetricName=?MetricName) :- LogRule(60), match(?MetricName, "DiskSpace"), GetRepairHistory(?repairCount, 08:00:00),
	?repairCount < 4,
	CheckFolderSize("D:\SvcFab\Log\Traces", MaxFolderSizeGB=20),
	DeleteFiles("D:\SvcFab\Log\Traces", SortOrder=Ascending, MaxFilesToDelete=10, RecurseSubdirectories=true, SearchPattern="*.etl").
```

I've then created 20 x 1GB test files in the C:\fabric_observer_logs directory (named for example `1gb.1.test`) these were created with `fsutil` and are 1GB in size each.

I was expecting (perhaps after 2 hours) for Fabric Healer to delete the oldest 10 of my files in this directory, and then 2 hours after this delete another 10 of them, but it's been about 6 hours and I'm not seeing anything occuring.

fh_operations_telemetry.log:

```
{"EventName":"OperationalEvent","TaskName":"FabricHealer","EventRunInterval":"1.00:00:00","SFRuntimeVersion":"9.1.1583.9590","ClusterId":"9bb1353a-355d-41e3-98c9-e990adf9c018","ClusterType":"SFRP","NodeNameHash":"e0f9d96061c0cdd36a0659711ed0768625f395a60a295c42242bbeb694f415eb","FHVersion":"1.2.0","UpTime":"00:00:00.1368888","Timestamp":"2023-04-06T15:27:02.0266516Z","OS":"Windows","EnabledRepairCount":2,"TotalRepairAttempts":0,"SuccessfulRepairs":0,"FailedRepairs":0}
```

RepairData.log

```
2023-04-06 09:55:44.9428--INFO--Detected Fabric node _SFnode0_0 is in Warning.
Machine repair target specified. 11 logic rules found for Machine repair.
2023-04-06 14:10:37.6579--INFO--Detected Fabric node _node5_1 is in Warning.
Machine repair target specified. 11 logic rules found for Machine repair.
```

I am seeing Forbidden errors in TelemetryLogger.log, which I assume is it not being able to write to Log Analytics despite me checking the workspace Id and Shared Key values are correct. I assume though that this is likely unrelated to why my Disk mitigation seems to be not firing.

Grateful if you can advise of anything I am doing wrong.

Thanks,
Mark

**Comment by markwragg** [comment]
Btw I added the `LogRule()` parts to the mitigation config on the VM directly in the hopes of getting some additional debug logging, they weren't in the mitigation originally. After I did this I killed the process to get it to refresh its config.

**Comment by GitTorre** [comment]
Hi Mark,

A couple things.

1) FH needs to be running on each node in the cluster to perform disk-related repairs (folder cleaning).
2) FO needs to be configured to monitor the same folders you want FH to clean and warn when the threshold is breached. 
3) Disk repair needs to be enabled (you understand this already).

[Using MB size for testing purposes, note ```MaxFolderSizeMB```]

FO Config:

```Xml
 <!-- Folder size monitoring. If you want to monitor folder sizes, then enable this feature and provide full paths (and/or path env vars) as shown below. -->
    <Parameter Name="DiskObserverEnableFolderSizeMonitoring" DefaultValue="true" />
    <!-- Supply full path/threshold(in MB) pairs for Warning/Error level monitoring here. Format is "fullpath0, size0 | fullpath1, size1 | fullpath2, size2 ..." -->
    <Parameter Name="DiskObserverFolderPathsErrorThresholdsMb" DefaultValue="" />
    <Parameter Name="DiskObserverFolderPathsWarningThresholdsMb" DefaultValue="C:\fabric_observer_logs, 20 | D:\SomeFolder, 10 | %SOMEFULLPATH%, 4096 | %SOMEPATH%\SubFolder, 1024" />
```
Note: The metric fact that FO supplies is not related to Disk Space, it is Folder Size (MB), which FH converts to a fact named ```FolderSizeMB```. The rule you are using is looking for DiskSpace, so the rule that is actually running (in the sample logic program provided) is this one: 

```Prolog
## Constrain on folder size Error or Warning code.
Mitigate(ErrorCode=?ErrorCode) :- ?ErrorCode == "FO042" || ?ErrorCode == "FO043", GetRepairHistory(?repairCount, 08:00:00),
	?repairCount < 4,
	CheckFolderSize("C:\fabric_observer_logs", MaxFolderSizeMB=250),
	DeleteFiles("C:\fabric_observer_logs", SortOrder=Ascending, MaxFilesToDelete=5, RecurseSubdirectories=true, SearchPattern="*.dmp").
```

FH Config:

ApplicationManifest:
```Xml
<Parameter Name="EnableDiskRepair" DefaultValue="true" />
```

DiskRules.guan:

```Prolog
Mitigate(MetricName=?MetricName) :- match(?MetricName, "FolderSizeMB"), GetRepairHistory(?repairCount, 08:00:00), 
	?repairCount < 4,
	member(config(?X,?Y), [config("D:\SvcFab\Log\Traces", 50), config("C:\fabric_observer_logs", 20), config("E:\temp", 10)]), 
	CheckFolderSize(?X, MaxFolderSizeMB=?Y),
	DeleteFiles(?X, SortOrder=Ascending, MaxFilesToDelete=10, RecurseSubdirectories=true).
```

Depending upon the folder, you may have to run FH as system user. By default, this is not the case. 

```Xml
    </ConfigOverrides>
    <!-- Because of certain actions FabricHealer takes in a cluster (deleting files or restarting system service processes running at higher elevation than FabricHealer, for example), 
         FH may need to run as System user on Windows and root on Linux. Note that you should test FH out NOT running as System and validate that it works for your scenarios.
         For repairs that require SF API calls (RestartCodePackage, RestartReplica, etc) or simply schedule infrastructure repairs via RM (see MachineRules.guan), FH can (and therefore should) 
         run as default user (that is Network Service on Windows, sfappsuser on Linux).
    <Policies>
      <RunAsPolicy CodePackageRef="Code" UserRef="SystemUser" />
    </Policies> -->
  </ServiceManifestImport>
  <Principals>
    <Users>
      <!-- LocalSystem AccountType maps to System on Windows and root on Linux. -->
      <User Name="SystemUser" AccountType="LocalSystem" />
    </Users>
  </Principals>
</ApplicationManifest>
```

All of this said, there is a bug... :(. I just fixed it while writing up this reply. nodeName was null in ExecutorData, which caused a failure in scheduling the repair (FH code), which means execution will not take place... My apologies. 

That said, I hope the above information is useful.

**Comment by markwragg** [comment]
That is useful thank you! It was your number 2 item I had not done.  However does this not also mean the examples in the are misleading as they use the DiskSpace metric.

I assume I also need to wait for a new release with the fix before it will work?

**Comment by GitTorre** [comment]
Fix is in develop. Would you mind testing it out? (I did, but want to make sure you see it working).

**Comment by GitTorre** [comment]
> That is useful thank you! It was your number 2 item I had not done. However does this not also mean the examples in the are misleading as they use the DiskSpace metric.
> 
> I assume I also need to wait for a new release with the fix before it will work?

Not really, as if there is a disk space issue, it is totally reasonable that you want to start deleting files in some folders where you can handle that. The folder monitoring in DiskObserver came later than what FH supports, so it supports deleting files in each case.

**Comment by GitTorre** [comment]
This will ship in 1.2.1, so, yeah, you will need to wait or add the fix and build yourself. It is a simple change in DeleteFilesPredicateType.cs in the Repair\Guan folder. One line:

```C#
               // DiskRepairPolicy
                var diskRepairPolicy = new DiskRepairPolicy
                {
                    FolderPath = path,
                    NodeName = RepairData.NodeName, // <-- bug fix.
                    MaxNumberOfFilesToDelete = maxFilesToDelete,
                    FileAgeSortOrder = direction,
                    RecurseSubdirectories = recurseSubDirectories,
                    FileSearchPattern = !string.IsNullOrWhiteSpace(searchPattern) ? searchPattern : "*",
                    RepairId = id,
                    MaxTimePostRepairHealthCheck = maxTimePostRepairHealthCheck,
                    DoHealthChecks = doHealthChecks,
                    RepairAction = RepairActionType.DeleteFiles
                };
```

**Comment by GitTorre** [comment]
I will add new unit tests so this doesn't happen again. This is my fault for not catching this well before 1.2.0 shipped...

**Comment by GitTorre** [comment]
> Btw I added the `LogRule()` parts to the mitigation config on the VM directly in the hopes of getting some additional debug logging, they weren't in the mitigation originally. After I did this I killed the process to get it to refresh its config.

That logs the currently executing rule. You would see the full rule in telemetry and SFX.

I didn't quite understand the telemetry issue you mentioned. Did you not receive any telemetry?

**Comment by GitTorre** [comment]
1.2.1 will be released on Monday.

**Comment by markwragg** [comment]
Thanks Charles. It's a long public holiday weekend in the UK due to Easter, so I'll be able to give it a test on Tuesday.

> I didn't quite understand the telemetry issue you mentioned. Did you not receive any telemetry?

Yes I'm not seeing a custom log called FabricHealer get created in Log Analytics and I'm assuming thats because of the "Forbidden" errors esesentially indicating its rejecting my key, but its strange because I've checked the key and workspace ID parameters are correctly populated and the way I've populated them (in my deployment pipeline) is the same for Service Fabric Observer and they do work fine there (I get the FabricObserver_CL and ClusterObserver_CL custom logs be created and populated).

**Comment by GitTorre** [comment]
> Thanks Charles. It's a long public holiday weekend in the UK due to Easter, so I'll be able to give it a test on Tuesday.
> 
> > I didn't quite understand the telemetry issue you mentioned. Did you not receive any telemetry?
> 
> Yes I'm not seeing a custom log called FabricHealer get created in Log Analytics and I'm assuming thats because of the "Forbidden" errors esesentially indicating its rejecting my key, but its strange because I've checked the key and workspace ID parameters are correctly populated and the way I've populated them (in my deployment pipeline) is the same for Service Fabric Observer and they do work fine there (I get the FabricObserver_CL and ClusterObserver_CL custom logs be created and populated).

Fixed in 1.2.1.

**Comment by GitTorre** [comment]
Note that I added Disk repair documentation to Using.md. That should help clarify the feature.
